File: windows/src/PhoneBridge.Pairing/PakeExchange.cs
using Org.BouncyCastle.Crypto.Agreement.JPake;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Security;
using System.Security.Cryptography;

namespace PhoneBridge.Pairing;

internal sealed class PakeExchange : IDisposable
{
    private JPakeParticipant? _participant;
    private JPakeRound1Payload? _ownFirst, _peerFirst;
    private JPakeRound2Payload? _ownSecond;
    private JPakeRound3Payload? _ownThird;
    private BigInteger? _key;
    private readonly string _own, _peer;

    internal PakeExchange(string own, string peer, char[] code)
    {
        _own = own; _peer = peer;
        _participant = new JPakeParticipant(own, code, JPakePrimeOrderGroups.NIST_3072, new Sha256Digest(), new SecureRandom());
    }

    internal byte[] Round1()
    {
        _ownFirst ??= _participant!.CreateRound1PayloadToSend();
        var p = _ownFirst;
        return Join(U(p.Gx1,384), U(p.Gx2,384), U(p.KnowledgeProofForX1[0],384), U(p.KnowledgeProofForX1[1],32),
            U(p.KnowledgeProofForX2[0],384), U(p.KnowledgeProofForX2[1],32));
    }

    internal void AcceptRound1(byte[] b)
    {
        _ = Round1();
        var p = new JPakeRound1Payload(_peer, Element(b,0), Element(b,384,true),
            [Element(b,768), Scalar(b,1152)], [Element(b,1184), Scalar(b,1568)]);
        _participant!.ValidateRound1PayloadReceived(p); _peerFirst = p;
    }

    internal byte[] Round2()
    {
        _ownSecond ??= _participant!.CreateRound2PayloadToSend();
        return Join(U(_ownSecond.A,384), U(_ownSecond.KnowledgeProofForX2s[0],384), U(_ownSecond.KnowledgeProofForX2s[1],32));
    }

    internal void AcceptRound2(byte[] b)
    {
        _ = Round2();
        _participant!.ValidateRound2PayloadReceived(new JPakeRound2Payload(_peer, Element(b,0,true), [Element(b,384), Scalar(b,768)]));
    }

    internal byte[] Round3()
    {
        _key ??= _participant!.CalculateKeyingMaterial();
        _ownThird ??= _participant!.CreateRound3PayloadToSend(_key);
        return S(_ownThird.MacTag);
    }

    internal void AcceptRound3(byte[] b)
    {
        _ = Round3();
        var expected = JPakeUtilities.CalculateMacTag(_peer, _own, _peerFirst!.Gx1, _peerFirst.Gx2,
            _ownFirst!.Gx1, _ownFirst.Gx2, _key!, new Sha256Digest());
        var tag = S(expected);
        try { if (!CryptographicOperations.FixedTimeEquals(tag,b)) throw new PairingProtocolException(); }
        finally { CryptographicOperations.ZeroMemory(tag); }
        _participant!.ValidateRound3PayloadReceived(new JPakeRound3Payload(_peer,new BigInteger(b)),_key!);
    }

    internal byte[] ExportKey() => U(_key ?? throw new PairingProtocolException(),384);
    private static BigInteger Element(byte[] b, int offset, bool notOne = false)
    {
        var v = new BigInteger(1,b.AsSpan(offset,384));
        if (v.SignValue <= 0 || v.CompareTo(JPakePrimeOrderGroups.NIST_3072.P) >= 0 || (notOne && v.Equals(BigInteger.One)))
            throw new PairingProtocolException();
        return v;
    }
    private static BigInteger Scalar(byte[] b, int offset)
    {
        var v = new BigInteger(1,b.AsSpan(offset,32));
        if (v.CompareTo(JPakePrimeOrderGroups.NIST_3072.Q) >= 0) throw new PairingProtocolException();
        return v;
    }
    private static byte[] U(BigInteger v,int length)
    {
        byte[] bytes = v.ToByteArrayUnsigned();
        if (v.SignValue < 0 || bytes.Length > length) throw new PairingProtocolException();
        var output = new byte[length]; bytes.CopyTo(output,length-bytes.Length);
        CryptographicOperations.ZeroMemory(bytes); return output;
    }
    private static byte[] S(BigInteger v)
    {
        byte[] bytes = v.ToByteArray();
        if (bytes.Length > 32) throw new PairingProtocolException();
        var output = Enumerable.Repeat(v.SignValue < 0 ? (byte)255 : (byte)0,32).ToArray();
        bytes.CopyTo(output,32-bytes.Length); return output;
    }
    private static byte[] Join(params byte[][] pieces) => pieces.SelectMany(p => p).ToArray();
    public void Dispose()
    {
        // BC exposes no abort/zeroize API; release references, never reuse a failed participant.
        _participant = null; _key = null; _ownFirst = _peerFirst = null; _ownSecond = null; _ownThird = null;
    }
}

File: android/pairing-core/src/main/kotlin/org/phonebridge/pairing/PakeExchange.kt
package org.phonebridge.pairing

import java.math.BigInteger
import java.security.SecureRandom
import org.bouncycastle.crypto.agreement.jpake.*
import org.bouncycastle.crypto.digests.SHA256Digest

internal class PakeExchange(private val own: String, private val peer: String, code: CharArray) : AutoCloseable {
    private var participant: JPAKEParticipant? = JPAKEParticipant(own,code,JPAKEPrimeOrderGroups.NIST_3072,SHA256Digest(),SecureRandom())
    private var first: JPAKERound1Payload? = null
    private var second: JPAKERound2Payload? = null
    private var third: JPAKERound3Payload? = null
    private var key: BigInteger? = null

    fun round1(): ByteArray {
        if (first == null) first = participant!!.createRound1PayloadToSend()
        val p = first!!
        return u(p.gx1,384)+u(p.gx2,384)+u(p.knowledgeProofForX1[0],384)+u(p.knowledgeProofForX1[1],32)+
            u(p.knowledgeProofForX2[0],384)+u(p.knowledgeProofForX2[1],32)
    }
    fun acceptRound1(b: ByteArray) {
        round1()
        participant!!.validateRound1PayloadReceived(JPAKERound1Payload(peer,element(b,0),element(b,384,true),
            arrayOf(element(b,768),scalar(b,1152)),arrayOf(element(b,1184),scalar(b,1568))))
    }
    fun round2(): ByteArray {
        if (second == null) second = participant!!.createRound2PayloadToSend()
        val p = second!!
        return u(p.a,384)+u(p.knowledgeProofForX2s[0],384)+u(p.knowledgeProofForX2s[1],32)
    }
    fun acceptRound2(b: ByteArray) {
        round2()
        participant!!.validateRound2PayloadReceived(JPAKERound2Payload(peer,element(b,0,true),arrayOf(element(b,384),scalar(b,768))))
    }
    fun round3(): ByteArray {
        if (key == null) key = participant!!.calculateKeyingMaterial()
        if (third == null) third = participant!!.createRound3PayloadToSend(key!!)
        return signed(third!!.macTag)
    }
    fun acceptRound3(b: ByteArray) {
        round3()
        // Pinned BC Java 1.86 validates the fixed-width MAC in constant time.
        participant!!.validateRound3PayloadReceived(JPAKERound3Payload(peer,BigInteger(b)),key!!)
    }
    fun exportKey(): ByteArray = u(key ?: throw PairingProtocolException(),384)

    private fun element(b: ByteArray, offset: Int, notOne: Boolean = false): BigInteger {
        val v = BigInteger(1,b.copyOfRange(offset,offset+384))
        if (v.signum() <= 0 || v >= JPAKEPrimeOrderGroups.NIST_3072.p || (notOne && v == BigInteger.ONE))
            throw PairingProtocolException()
        return v
    }
    private fun scalar(b: ByteArray, offset: Int): BigInteger {
        val v = BigInteger(1,b.copyOfRange(offset,offset+32))
        if (v >= JPAKEPrimeOrderGroups.NIST_3072.q) throw PairingProtocolException()
        return v
    }
    private fun u(v: BigInteger, size: Int): ByteArray {
        val bytes = v.toByteArray()
        val offset = if (bytes.size > 1 && bytes[0] == 0.toByte()) 1 else 0
        if (v.signum() < 0 || bytes.size-offset > size) throw PairingProtocolException()
        val out = ByteArray(size); bytes.copyInto(out,size-bytes.size+offset,offset); bytes.fill(0); return out
    }
    private fun signed(v: BigInteger): ByteArray {
        val bytes = v.toByteArray()
        if (bytes.size > 32) throw PairingProtocolException()
        val out = ByteArray(32) { if (v.signum() < 0) (-1).toByte() else 0 }
        bytes.copyInto(out,32-bytes.size); return out
    }
    override fun close() {
        // The library has no abort/zeroize API. Failed instances are never reused.
        participant = null; first = null; second = null; third = null; key = null
    }
}
